pkg/debug/common/common.go

```diff
@@ -0,0 +1,245 @@
+	if tail < 0 {
+		podLogOpts = corev1.PodLogOptions{}
+	} else {
+		podLogOpts = corev1.PodLogOptions{
+			TailLines: func(i int64) *int64 { return &i }(tail),
+		}
+	}
+	req := c.CoreV1().Pods(ns).GetLogs(pod, &podLogOpts)
+	podLogs, err := req.Stream()
+	if err != nil {
+		return "", fmt.Errorf("failed to open stream: %s", err.Error())
+	}
+	defer podLogs.Close()
+
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, podLogs)
+	if err != nil {
+		return "", fmt.Errorf("failed to copy information from podLogs to buf: %s", err.Error())
+	}
+	return buf.String(), nil
+}
+
+// Exec executes certain command and returns the result
+func Exec(pod string, ns string, cmd string, c *kubernetes.Clientset) (string, error) {
+	req := c.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(pod).
+		Namespace(ns).
+		SubResource("exec")
+
+	req.VersionedParams(&corev1.PodExecOptions{
+		Command: []string{"/bin/sh", "-c", cmd},
+		Stdin:   false,
+		Stdout:  true,
+		Stderr:  true,
+		TTY:     false,
+	}, kubectlscheme.ParameterCodec)
+
+	var stdout, stderr bytes.Buffer
+	exec, err := remotecommand.NewSPDYExecutor(config.GetConfigOrDie(), "POST", req.URL())
+	if err != nil {
+		return "", fmt.Errorf("error in creating NewSPDYExecutor: %s", err.Error())
+	}
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return "", fmt.Errorf("error in creating StreamOptions: %s", err.Error())
+	}
+	if stderr.String() != "" {
+		return stdout.String(), fmt.Errorf(stderr.String())
+	}
+	return stdout.String(), nil
+}
+
+// GetChaos returns the chaos that will do type-assertion
+func GetChaos(ctx context.Context, chaosType string, chaosName string, ns string, c client.Client) (runtime.Object, error) {
+	// get podName
+	chaosType = upperCaseChaos(strings.ToLower(chaosType))
+	allKinds := v1alpha1.AllKinds()
+	chaos := allKinds[chaosType].Chaos
+	objectKey := client.ObjectKey{
+		Namespace: ns,
+		Name:      chaosName,
+	}
+	if err := c.Get(ctx, objectKey, chaos); err != nil {
+		return nil, fmt.Errorf("failed to get chaos %s: %s", chaosName, err.Error())
+	}
+	return chaos, nil
+}
+
+// GetPods returns pod list and corresponding chaos daemon
+func GetPods(ctx context.Context, status v1alpha1.ChaosStatus, selector v1alpha1.SelectorSpec, c client.Client) ([]v1.Pod, []v1.Pod, error) {
+	// get podName
+	failedMessage := status.FailedMessage
+	if failedMessage != "" {
+		return nil, nil, fmt.Errorf("chaos failed with: %s", failedMessage)
+	}
+
+	phase := status.Experiment.Phase
+	nextStart := status.Scheduler.NextStart
+
+	if phase == "Waiting" {
+		waitTime := nextStart.Sub(time.Now())
+		fmt.Printf("Waiting for chaos to start, in %s\n", waitTime)
+		time.Sleep(waitTime)
+	}
+
+	// TODO: failed and maybe not appropirate to create manager here, to get client.Reader
+	// So could not parse fieldSelector for now
+	pods, err := utils.SelectPods(ctx, c, nil, selector)
```

Comment:
Failed and maybe not appropirate to create a manager here, to get client.Reader. So could not parse fieldSelector for now. Any suggestions?

Reply:
Thx for the suggestions! It seems still uses `client` from `controller-runtime` in the links. Is there someplace I misunderstood?

Reply:
Ouch, I misremembered about that.  It seems chaos-dashboard does not using what I say.

TLDR, you could use `utils.SelectPods(ctx, c, nil, selector)` without any issue.

When we get `client.Client` from `Manager`(It's a struct of controller-manager), like what `chaos-dashboard` dose in [here](https://github.com/chaos-mesh/chaos-mesh/blob/6af3d9f70b760a50867425ff1be4d01ca3041c81/pkg/collector/server.go#L95), this client will contains a cache(index), this causing problem with fieldSelector.
When you dig into the constructor of `Manager` you could find out that `apiReader` is actually the normal client, `client` is a combination of `apiReader` and a `cache`, [ref](https://github.com/kubernetes-sigs/controller-runtime/blob/114431a4df156146338f5bb190af0a0664144b24/pkg/manager/manager.go#L380). And you also could find out that `client.Client` is a superset for `client.Reader`.

So the way you create `client.Client` will make a client without cache, it will NOT cause issues about fieldSelector. 🎉

Reply:
That's it. I think I understand for now. Thank you for the explanation!! 

Actually there is a tiny issue, since if we need to select with fieldSelectors, current version would forcibly use `client.Reader`.
https://github.com/chaos-mesh/chaos-mesh/blob/6afb841b02df35a03bbca84419e70cb95eb8c8e1/pkg/utils/selector.go#L140-L150

Considering [another PR](https://github.com/chaos-mesh/chaos-mesh/pull/1112/files#diff-8a6b3b234279ff422c0e1c449141a18d54d29f47d383ab2d549a2a0788503ac8R95) encounter the same problem, we could add another case - when `reader` is `nil`, still `list` from `client`

Reply:
How about using `utils.SelectPods(ctx, c, c, selector)`, as you known `client.Client` is a superset for `client.Reader`.

Reply:
Ah yeah that's a good idea! thx🎉